src/Conversion/ONNXToKrnl/Math/Elementwise.cpp

```diff
@@ -1385,6 +1388,122 @@ static LogicalResult getVariadicSimdCodeFullyFlattened(
+    Location loc, Operation *op, Type elementType,
+    ArrayRef<Value> scalarOperands);
+
+// Variadic template to iterate all the fusible Ops
+template <typename T>
+bool enqueueFusedOpImpl(Operation *op, SmallVector<Operation *, 2> &fusibleOps,
+    SmallVector<EmitScalarFunc, 2> &fuseEmitFunctions) {
+  if (isa<T>(op)) {
+    fusibleOps.emplace_back(op);
+    fuseEmitFunctions.emplace_back(emitScalarOpFor<T>);
+    return true;
+  } else {
+    return false;
+  }
+}
+
+template <typename T = void, class... Ts>
+bool enqueueFusedOp(Operation *op, SmallVector<Operation *, 2> &fusibleOps,
+    SmallVector<EmitScalarFunc, 2> &fuseEmitFunctions);
+
+template <typename T, class... Ts>
+bool enqueueFusedOp(Operation *op, SmallVector<Operation *, 2> &fusibleOps,
+    SmallVector<EmitScalarFunc, 2> &fuseEmitFunctions) {
+  if (enqueueFusedOpImpl<T>(op, fusibleOps, fuseEmitFunctions)) {
+    return true;
+  } else {
+    return enqueueFusedOp<Ts...>(op, fusibleOps, fuseEmitFunctions);
+  }
+}
+
+template <>
+bool enqueueFusedOp(Operation *op, SmallVector<Operation *, 2> &fusibleOps,
+    SmallVector<EmitScalarFunc, 2> &fuseEmitFunctions) {
+  return false;
+}
+
+bool enqueueFusedOp(Operation *op, SmallVector<Operation *, 2> &fusibleOps,
+    SmallVector<EmitScalarFunc, 2> &fuseEmitFunctions) {
+
+  // Notice: Though ClipOp is classified as unary element op in this file,
+  // ClipOp requires one required input and two optional input
+
+  return enqueueFusedOp<ONNXCastOp, ONNXSinhOp, ONNXCoshOp, ONNXSigmoidOp,
+      ONNXHardSigmoidOp, ONNXEluOp, ONNXReluOp, ONNXLeakyReluOp, ONNXPReluOp,
+      ONNXSeluOp, ONNXReciprocalOp, ONNXSoftplusOp, ONNXSoftsignOp, ONNXSignOp,
+      ONNXMaxOp, ONNXMinOp, ONNXNegOp, ONNXLessOp, ONNXLessOrEqualOp,
+      ONNXGreaterOp, ONNXGreaterOrEqualOp, ONNXEqualOp, ONNXNotOp, ONNXModOp,
+      ONNXMeanOp, ONNXRoundOp, ONNXDequantizeLinearOp, ONNXSqrtOp>(
+      op, fusibleOps, fuseEmitFunctions);
+}
+
+// This function starts for elementwise operation, op.
+// A successor op (user) is fusible if it's the only user and an unary
+// elementwise Op. The Op and its EmitScalarOpFor<T> are recorded into
+// the vector.
+static bool findFusibleUnaryOps(Operation *op,
+    SmallVector<Operation *, 2> &fusibleOps,
+    SmallVector<EmitScalarFunc, 2> &fuseEmitFunctions) {
+  Operation *currentProducer = op;
+  while (currentProducer->hasOneUse()) {
+    // Check the users is an unary elementwise op
+    // The right solution, I think, is to define EmitScalarOpFor as
+    // an interface of unary elementwise Ops.
+    // In the draft PR, variadic template is used to iterate through
+    // the possible ONNX Ops.
+    Operation *user = *currentProducer->getUsers().begin();
+    if (!enqueueFusedOp(user, fusibleOps, fuseEmitFunctions))
+      break;
+
+    currentProducer = user;
+  }
+
+  if (fusibleOps.size() > 0)
+    LLVM_DEBUG(
+        { llvm::dbgs() << "unary op fused: " << fusibleOps.size() << "\n"; });
```

Comment:
Uhm, is it seriously checking the condition for LLVM_DEBUG here?

Reply:
Condition is to reduce the debug info printout.